### PR TITLE
Bail on invalid folder

### PR DESCRIPTION
**Pull request by kunagpal**
Addresses https://github.com/postmanlabs/newman/issues/1061

### CODE REVIEW

File: lib/runner/index.js [hunk 1]
```diff
@@ -39,13 +39,21 @@ _.assign(Runner.prototype, {
...
+        var runnableItems,
+            runOptions = _.merge(_.omit(options, ['environment', 'globals', 'data']), this.options.run);
 
         callback = backpack.normalise(callback);
         !_.isObject(options) && (options = {});
 
+        runnableItems = Runner.extractRunnableItems(collection, options.entrypoint);
+
+        // Bail out only if: abortOnError is set and the returned entrypoint is invalid
+        if (options.abortOnError && !runnableItems.entrypoint) {
```
Comment (shamasis) [inline lib/runner/index.js]: `extractRunnableItems` returns array, then what is this check `!runnableItems.entrypoint`?

Comment (kunagpal) [inline lib/runner/index.js]: The return structure has changed from an array to an object with the resolved entry point and extracted run items:
```json
{
    entryPoint: ItemGroup
    items: []
}
```

File: lib/runner/index.js [hunk 2]
```diff
@@ -39,13 +39,21 @@ _.assign(Runner.prototype, {
...
+            return callback(new Error('A valid folder id/name is required'));
```
Comment (kunagpal) [inline lib/runner/index.js]: The default evaluation on this case is the same as before (the collection)

File: lib/runner/index.js (hunk 1 above)
Comment (shamasis) [inline lib/runner/index.js]: Instead of this, why not make `extractRunnables` accept callback with multiple params

Comment (kunagpal) [inline lib/runner/index.js]: On it 👍

File: lib/runner/index.js [hunk 3]
```diff
@@ -39,19 +40,31 @@ _.assign(Runner.prototype, {
...
-            proxies: options.proxies
-        }, runOptions)));
+        Runner.extractRunnableItems(collection, options.entrypoint, function (err, runnableItems) {
+            if (err) { return callback(err); }
+            // The condition below will probably never be satisfied
+            if (!runnableItems) { return callback(new Error('Error fetching run items')); }
+
+            // Bail out only if: abortOnError is set and the returned entrypoint is invalid
+            if (options.abortOnError && !runnableItems.entrypoint) {
+                return callback(new Error('A valid folder id/name is required'));
```
Comment (kunagpal) [inline lib/runner/index.js]: Unable to find a folder or request: 'options.entrypoint'

File: lib/runner/index.js [hunk 4]
```diff
@@ -39,19 +40,31 @@ _.assign(Runner.prototype, {
...
 
-        return callback(null, (new Run({
-            items: Runner.extractRunnableItems(collection, options.entrypoint),
-            data: Runner.normaliseIterationData(options.data, options.iterationCount),
-            environment: options.environment,
-            globals: _.has(options, 'globals') ? options.globals : this.options.globals,
-            certificates: options.certificates,
-            proxies: options.proxies
-        }, runOptions)));
+        Runner.extractRunnableItems(collection, options.entrypoint, function (err, runnableItems) {
```
Comment (shamasis) [inline lib/runner/index.js]: make callback of extract runnables pass two params.

File: lib/runner/index.js [hunk 5]
```diff
@@ -90,26 +103,43 @@ _.assign(Runner, {
...
+     * @param {?Collection} collection
+     * @param {?String} groupId
+     * @param {Function} callback
+     * @returns {Object}
+     * @todo Add support for accepting groupId as an array of ids/names
      */
-    extractRunnableItems: function (collection, groupId) {
-        var items = [],
+    extractRunnableItems: function (collection, groupId, callback) {
+        _.isFunction(groupId) && (callback = groupId, groupId = null);
```
Comment (shamasis) [inline lib/runner/index.js]: polymorphism here is an overkill

File: lib/runner/index.js [hunk 6]
```diff
@@ -90,26 +103,43 @@ _.assign(Runner, {
...
+        if (!Collection.isCollection(collection)) {
+            return callback(new Error('A valid collection SDK instance is required'));
+        }
+
+        var idMatch,
+            nameMatch,
+            items = [],
             group = !groupId && collection; // if a group id is unspecified, the collection is our group
 
         // if group id is provided, we try and check if we can find it in the collection
```
Comment (shamasis) [inline lib/runner/index.js]: comment needs to be negated. :-/

File: lib/runner/index.js [hunk 7]
```diff
@@ -90,26 +103,43 @@ _.assign(Runner, {
...
-        !group && collection.forEachItemGroup(function (itemGroup) {
-            // @todo: fix this. the check by id should all finish before doing check by name
-            (itemGroup.id === groupId || itemGroup.name === groupId) && (group = itemGroup);
-        });
+        if (!group) {
+            collection && collection.forEachItemGroup && collection.forEachItemGroup(function (itemGroup) {
```
Comment (shamasis) [inline lib/runner/index.js]: redundant if collection is typechecked earlier

File: lib/runner/index.js (hunk 4 above)
Comment (kunagpal) [reply lib/runner/index.js]: Addressed

File: lib/runner/index.js (hunk 5 above)
Comment (kunagpal) [reply lib/runner/index.js]: Addressed

File: lib/runner/index.js (hunk 6 above)
Comment (kunagpal) [reply lib/runner/index.js]: Addressed

File: lib/runner/index.js (hunk 7 above)
Comment (kunagpal) [reply lib/runner/index.js]: Addressed

**Comment by kamalaknn** [comment]
Superceded by #427 

Adds improvements on top of this.

**Pull request closed without merging**